src/utils/parallelPaginatedRequest.js

```diff
@@ -0,0 +1,73 @@
+import axios from 'axios';
+
+async function fetchPage(apiUrl, requestConfig, page, pageSize, maxRetries) {
+  let attempt = 0;
+  let data = null;
+
+  while (attempt < maxRetries) {
+    attempt++;
+
+    try {
+      const response = await axios.get(`${apiUrl}&page=${page}&page_size=${pageSize}`, requestConfig);
+
+      if (response.status !== 200) {
+        throw new Error(`HTTP error ${response.status}`);
+      }
+
+      data = response.data.data;
+      break;
+    } catch (error) {
+      console.log(`Error fetching page ${page} (attempt ${attempt}):`, error.message);
+    }
+  }
+
+  if (attempt >= maxRetries) { // just logging in error, this place can be used to trigger a callback on maxRetries exhausted per page
+    console.error(`Failed to fetch page ${page} after ${maxRetries} attempts.`);
+  }
+
+  return data;
+}
+
+async function parallelPaginatedRequest(apiUrl, requestConfig = {}, { itemsPerPage = 150, maxRetries = 3, concurrencyLimit = 5, onPageFetch = null } = {}) {
+
+  const firstPageData = await fetchPage(apiUrl, requestConfig, 1, itemsPerPage, maxRetries);
+
+  if (!firstPageData) {
+    throw new Error('Failed to fetch the first page. Aborting.');
+  }
+
+  onPageFetch?.(firstPageData);
+
+  const totalItems = firstPageData.count;
+  let allItems = [...firstPageData.results];
+  const pagesToFetch = Array.from({ length: Math.ceil(totalItems / itemsPerPage) - 1 }, (_, i) => i + 2); // Exclude page 1
+  const activeFetches = new Set();
+
+  const processPage = async (page) => {
+    const data = await fetchPage(apiUrl, requestConfig, page, itemsPerPage, maxRetries);
+    if (!!data) {
+      onPageFetch?.(data);
+      allItems = allItems.concat(data.results);
+    }
+  };
+
+  for (const page of pagesToFetch) {
+    while (activeFetches.size >= concurrencyLimit) {
+      await Promise.race([...activeFetches]);
```

Comment:
Hmm I don't understand why we are using Promise.race here 🤔 The nature of this feature is not a race at all. We not only care about the response of the fastest request. We want to call 5 promises in parallel and everytime one responds just add another one, like a pool. It seems like everytime the race finishes, we will iterate in the while  calling the Promise.race again with the 4 requests that didn't finish, causing doing the same requests several times (?) Could you explain a little the reasoning behind the race here please.

Reply:
That's a typo, I thought I was fixed it, it should be an if statement

Regarding the Promise.race, it finishes when one of the promises of the set is processed, meaning it succeed or failed, so that result is taken and Promise.race if finished also, as far as I understand it only checks promises statuses so there is no request duplication